File: advanced-programing-lang/week-2/part-2/factorial.js
function doFactorialCounter() {
    let count = 0; // Enclosed variable for closure

    return function factorial(n) {
        count++;
        if (n < 0) return 0;
        
        let result = 1;
        for (let i = 1; i <= n; i++) {
            result *= i;
        }
        return { result, count };
    };
}

// Usage
const factorialGen = doFactorialCounter();
const res1 = factorialGen(5);
console.log(`Result: ${res1.result}, Called: ${res1.count} times`); // Result: 120, Called: 1 times

File: advanced-programing-lang/week-2/part-2/factorial.py
def do_factorial_counter():
    count = 0  # Enclosed variable for closure
    
    def factorial(n: int) -> int:
        nonlocal count
        count += 1
        
        if n < 0: 
            return 0
        result = 1
        for i in range(1, n + 1):
            result *= i
        return result, count
        
    return factorial

# Usage
fact_gen = do_factorial_counter()
res1, c1 = fact_gen(5)
print(f"Result: {res1}, Called: {c1} times") # Result: 120, Called: 1 times